### PR TITLE
aws health checks

### PR DESCRIPTION
**Pull request by philbrookes**
First pass at AWS health checks in DNS Operator.

Some uses-cases not fully covered, and no smart backoff, or e2e tests yet.

Related PR for kuadrant-operator:
https://github.com/Kuadrant/kuadrant-operator/pull/543

### CODE REVIEW

File: internal/controller/dnsrecord_controller.go
Comment (maleck13) [inline internal/controller/dnsrecord_controller.go:136]: might be a bit noisey?

File: internal/controller/dnsrecord_controller.go
Comment (maleck13) [inline internal/controller/dnsrecord_controller.go:139]: how valuable are these logs? Seems like they would also get quite noisy?

File: internal/controller/dnsrecord_controller.go
Comment (maleck13) [inline internal/controller/dnsrecord_controller.go:257]: this log looks inaccurate ?

File: internal/controller/dnsrecord_healthchecks.go
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:37]: having a reconcile for deletion and a reconcile for adding / updating feels a bit odd? Why can't it just be done in the one reconcile?

File: internal/controller/dnsrecord_healthchecks.go
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:68]: intresting so will we not have any health checks if the using an AWS LB as it uses hostnames?

File: internal/controller/dnsrecord_controller.go
Comment (philbrookes) [reply internal/controller/dnsrecord_controller.go:136]: these were primarily for debugging, I will remove a bunch of them now.

File: internal/controller/dnsrecord_healthchecks.go
Comment (philbrookes) [reply internal/controller/dnsrecord_healthchecks.go:68]: ah good point, I need to also capture when it's an external CNAME.

File: internal/controller/dnsrecord_healthchecks.go
Comment (philbrookes) [reply internal/controller/dnsrecord_healthchecks.go:37]: this was brought in from the old MGC implementation, I can change it

Comment (maleck13) [reply internal/controller/dnsrecord_healthchecks.go:37]: well doesn't block merging, but just seems a bit cumbersome

File: internal/controller/dnsrecord_healthchecks.go
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:114]: also noisy?

File: internal/controller/dnsrecord_healthchecks.go
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:126]: what if this was fmt.Sprintf("all %v probes synced successfully",len(results))

then lower down you prob don't need the if (len(results) > 0 ) ?

File: internal/provider/endpointUtils.go
Comment (maleck13) [inline internal/provider/endpointUtils.go:12]: is there ever more than 1 target?

File: internal/controller/dnsrecord_healthchecks.go
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:190]: should we be ignoring the error here?

File: internal/provider/aws/health.go
Comment (maleck13) [inline internal/provider/aws/health.go:46]: should we log this here or just add it as context to the returned error? I imagine it will get logged here and then back at the dnsrecord reconciler?

File: internal/provider/aws/health.go
Comment (maleck13) [inline internal/provider/aws/health.go:68]: what is a `HealthCheckResult` in this context? Is the result of the update/create or is it whether the health check is failing / passing?

File: internal/provider/aws/health.go
Comment (maleck13) [reply internal/provider/aws/health.go:52]: I guess the client might be the place to do that?

File: internal/provider/health.go
Comment (maleck13) [inline internal/provider/health.go:62]: These fake and cached reconcilers, maybe would make sense in a different file

**Review by maleck13 — COMMENTED** [review]
few questions and comments. I realise there is some pressure to get this merged. But lets see what tests we can get in over the next day or so WDYT?

File: internal/provider/aws/health.go
Comment (philbrookes) [reply internal/provider/aws/health.go:62]: yeah, it's the result of the API call.

File: internal/controller/dnsrecord_controller.go
Comment (maleck13) [inline internal/controller/dnsrecord_controller.go:275]:
```suggestion
```

File: internal/controller/dnsrecord_healthchecks.go
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:40]: looks like redefining the var here?

File: internal/controller/dnsrecord_healthchecks.go
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:57]: whats this for?

File: internal/controller/dnsrecord_healthchecks.go
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:94]: I think this whole code block need a bit of re-working

Comment (maleck13) [reply internal/controller/dnsrecord_healthchecks.go:87]: So this is better but WDYT
```
for _, endpoint := range dnsRecord.Spec.Endpoints {
			addresses := provider.GetExternalAddresses(endpoint, dnsRecord)
			for _, address := range addresses {
				endpointId, err := idForEndpoint(dnsRecord, dnsEndpoint, address)
				if err != nil {
					return err
				}
				probeStatus := r.getProbeStatus(address, dnsRecord)
				// trigger delete
				if config == nil{
					log.Log.Info("deleting health checks for","endpoint ", endpoint)
					if probeStatus == nil {
						continue
					}
					result, err := healthCheckReconciler.Delete(ctx, endpoint, probeStatus)
					if err != nil {
						return err
					}

					results = append(results, result)
					continue
				}
				spec := provider.HealthCheckSpec{
					Id:               endpointId,
					Name:             fmt.Sprintf("%s-%s-%s", *dnsRecord.Spec.RootHost, dnsEndpoint.DNSName, address),
					Host:             dnsRecord.Spec.RootHost,
					Path:             config.Endpoint,
					Port:             config.Port,
					Protocol:         config.Protocol,
					FailureThreshold: config.FailureThreshold,
				}
				result, err := healthCheckReconciler.Reconcile(ctx, spec, dnsEndpoint, probeStatus, address)
				if err != nil {
					return err
				}
				results = append(results, result)
			}
		}
	

	result := r.reconcileHealthCheckStatus(results, dnsRecord)
	return result
```

Comment (maleck13) [reply internal/controller/dnsrecord_healthchecks.go:87]: basically drop the repeated loop in the code.

File: internal/controller/dnsrecord_healthchecks.go
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:113]: worth using ```meta.FindStatusCondition``` ?

File: internal/controller/dnsrecord_healthchecks.go
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:171]: seems a bit overkill ?

File: internal/controller/dnsrecord_healthchecks.go
Comment (philbrookes) [reply internal/controller/dnsrecord_healthchecks.go:113]: oh cool, I didn't know this existed.

File: internal/controller/dnsrecord_healthchecks.go
Comment (philbrookes) [reply internal/controller/dnsrecord_healthchecks.go:174]: I think this is just making it a little more legible and compact to convert stuff to pointers, does it need changing here?

**Review by maleck13 — COMMENTED** [review]
There being not tests here is really not ideal. We need to make sure we get these in place @philbrookes can you  link to an issue 

@mikenairn In order to unblock moving kuadarant operator back to main of dns operator, I think we can make an exception for this knowing the focus will be on adding tests before we do the next release of DNS operator?

**Comment by philbrookes** [comment]
> There being not tests here is really not ideal. We need to make sure we get these in place @philbrookes can you link to an issue
> 
> @mikenairn In order to unblock moving kuadarant operator back to main of dns operator, I think we can make an exception for this knowing the focus will be on adding tests before we do the next release of DNS operator?

My intention was to keep this issue open until the tests are added and the logic is firmed up more. https://github.com/Kuadrant/dns-operator/issues/44

**Review by maleck13 — APPROVED** [review]

**Pull request merged**
